gradle-plugin/src/main/java/com/squareup/anvil/plugin/AnvilPlugin.kt

```diff
@@ -1,60 +1,114 @@
-  }
+  override fun apply(target: Project) {
+    val extension = target.extensions.create("anvil", AnvilExtension::class.java)
 
-  override fun getCompilerPluginId(): String = "com.squareup.anvil.compiler"
+    // Create a configuration for collecting CodeGenerator dependencies. We need to create all
+    // configurations eagerly and cannot wait for applyToCompilation(..) below, because this
+    // function is called in an afterEvaluate block by the Kotlin Gradle Plugin. That's too late
+    // to register the configurations.
+    val commonConfiguration = getConfiguration(target, buildType = "")
+
+    // anvilTest is the common test configuration similar to other configurations like kaptTest.
+    // Android build type specific unit test variants will extend this variant below. For the JVM
+    // this single variant is enough.
+    val testConfiguration = getConfiguration(target, buildType = "test").apply {
+      extendsFrom(commonConfiguration)
+    }
 
-  override fun getPluginArtifact(): SubpluginArtifact = SubpluginArtifact(
-    groupId = GROUP,
-    artifactId = "compiler",
-    version = VERSION
-  )
+    agpPlugins.forEach { agpPlugin ->
+      target.pluginManager.withPlugin(agpPlugin) {
+        // This is the common android test variant, similar to anvilTest above.
+        val androidTestVariant = getConfiguration(target, buildType = "androidTest").apply {
+          extendsFrom(commonConfiguration)
+        }
+
+        target.androidVariantsConfigure { variant ->
+          // E.g. "debugAnvil", "testReleaseAnvil", ...
+          val configuration = getConfiguration(target, buildType = variant.name)
+
+          when (variant) {
+            is UnitTestVariant -> configuration.extendsFrom(testConfiguration)
+            is TestVariant -> configuration.extendsFrom(androidTestVariant)
+          }
+        }
+      }
+    }
+
+    target.afterEvaluate {
+      if (!extension.generateDaggerFactories.get() &&
+        extension.generateDaggerFactoriesOnly.get()
+      ) {
+        throw GradleException(
+          "You cannot set generateDaggerFactories to false and generateDaggerFactoriesOnly " +
+            "to true at the same time."
+        )
+      }
+    }
+  }
+
+  override fun isApplicable(kotlinCompilation: KotlinCompilation<*>): Boolean {
+    return when (kotlinCompilation.platformType) {
+      androidJvm, jvm -> true
+      else -> false
+    }
+  }
 
   override fun applyToCompilation(
     kotlinCompilation: KotlinCompilation<*>
   ): Provider<List<SubpluginOption>> {
-    val project = kotlinCompilation.target.project
+    val variant = Variant(kotlinCompilation)
+    val project = variant.project
+
+    // Make the kotlin compiler classpath extend our configurations to pick up our extra
+    // generators.
+    project.configurations.getByName(variant.compilerPluginClasspathName)
+      .extendsFrom(getConfiguration(project, variant.name))
```

Comment:
Here I'm making the connection between the Anvil configuration and the Kotlin compiler plugin classpath.